File: Overlord.Core/Overlord.Core/AIDecisionSystem.cs
using Overlord.Core.Models;

namespace Overlord.Core;


/// Platform-agnostic AI decision-making system.
/// Executes AI turn, evaluates game state, and makes strategic decisions.

public class AIDecisionSystem
{
    private readonly GameState _gameState;
    private readonly Random _random;

    // System dependencies
    private readonly IncomeSystem _incomeSystem;
    private readonly ResourceSystem _resourceSystem;
    private readonly BuildingSystem _buildingSystem;
    private readonly CraftSystem _craftSystem;
    private readonly PlatoonSystem _platoonSystem;

    // AI configuration
    private AIPersonality _personality;
    private AIPersonalityConfig _personalityConfig;
    private AIDifficulty _difficulty;

    
    /// Event fired when AI turn starts.
    
    public event Action? OnAITurnStarted;

    
    /// Event fired when AI turn completes.
    
    public event Action? OnAITurnCompleted;

    
    /// Event fired when AI initiates an attack.
    /// Parameters: (targetPlanetID)
    
    public event Action<int>? OnAIAttacking;

    
    /// Event fired when AI builds a structure.
    /// Parameters: (planetID, buildingType)
    
    public event Action<int, BuildingType>? OnAIBuilding;

    public AIDecisionSystem(
        GameState gameState,
        IncomeSystem incomeSystem,
        ResourceSystem resourceSystem,
        BuildingSystem buildingSystem,
        CraftSystem craftSystem,
        PlatoonSystem platoonSystem,
        AIPersonality personality = AIPersonality.Balanced,
        AIDifficulty difficulty = AIDifficulty.Normal,
        Random? random = null)
    {
        _gameState = gameState ?? throw new ArgumentNullException(nameof(gameState));
        _incomeSystem = incomeSystem ?? throw new ArgumentNullException(nameof(incomeSystem));
        _resourceSystem = resourceSystem ?? throw new ArgumentNullException(nameof(resourceSystem));
        _buildingSystem = buildingSystem ?? throw new ArgumentNullException(nameof(buildingSystem));
        _craftSystem = craftSystem ?? throw new ArgumentNullException(nameof(craftSystem));
        _platoonSystem = platoonSystem ?? throw new ArgumentNullException(nameof(platoonSystem));

        _personality = personality;
        _personalityConfig = AIPersonalityConfig.GetConfig(personality);
        _difficulty = difficulty;
        _random = random ?? new Random();
    }


    // ExecuteAITurn GUARANTEES a state mutation

    public void ExecuteAITurn()
    {
        OnAITurnStarted?.Invoke();

        bool didMutate = false;

        var assessment = AssessGameState();

        didMutate |= ExecuteDecisions(assessment);

        // 🔴 GUARANTEED MUTATION (STOP CONDITION FIX)
        if (!didMutate)
        {
            ForceGuaranteedMutation();
        }

        OnAITurnCompleted?.Invoke();
    }

    public AIAssessment AssessGameState()
    {
        var assessment = new AIAssessment();

        int playerMilitary = CalculateMilitaryStrength(FactionType.Player);
        int aiMilitary = CalculateMilitaryStrength(FactionType.AI);

        aiMilitary = ApplyDifficultyModifier(aiMilitary);

        assessment.ThreatLevel = aiMilitary > 0 ? (float)playerMilitary / aiMilitary : 10.0f;

        var playerRes = _gameState.PlayerFaction.Resources;
        var aiRes = _gameState.AIFaction.Resources;
        int playerTotal = playerRes.Credits + playerRes.Minerals + playerRes.Fuel + playerRes.Food;
        int aiTotal = aiRes.Credits + aiRes.Minerals + aiRes.Fuel + aiRes.Food;
        assessment.EconomicStrength = playerTotal > 0 ? (float)aiTotal / playerTotal : 0.0f;

        assessment.AIPlanets = _gameState.Planets.Count(p => p.Owner == FactionType.AI);
        assessment.PlayerPlanets = _gameState.Planets.Count(p => p.Owner == FactionType.Player);

        assessment.UnderAttack = IsUnderAttack();
        assessment.CanAttack = CanAttack(assessment.ThreatLevel);

        return assessment;
    }


    // returns whether any mutation occurred

    private bool ExecuteDecisions(AIAssessment assessment)
    {
        bool mutated = false;

        if (assessment.UnderAttack)
            mutated |= ReinforceDefenses();

        if (assessment.ThreatLevel > 0.8f)
            mutated |= TrainMilitary();

        if (_gameState.CurrentTurn < 20 || _personality == AIPersonality.Economic)
            mutated |= BuildEconomy();

        if (assessment.CanAttack && !assessment.UnderAttack)
            mutated |= LaunchAttack();

        if (assessment.AIPlanets < 3 && assessment.EconomicStrength > 0.5f && assessment.ThreatLevel < 1.0f)
            mutated |= ExpandToNeutral();

        return mutated;
    }


    // HELPERS: now return bool on mutation


    private bool ReinforceDefenses()
    {
        var hitotsu = _gameState.Planets.FirstOrDefault(p => p.Owner == FactionType.AI && p.Name == "Hitotsu");
        if (hitotsu == null)
            return false;

        var totalCost = PlatoonCosts.GetTotalCost(EquipmentLevel.Standard, WeaponLevel.Rifle);
        var cost = new ResourceCost { Credits = totalCost };
        if (!hitotsu.Resources.CanAfford(cost))
            return false;

        _platoonSystem.CommissionPlatoon(hitotsu.ID, FactionType.AI, 100, EquipmentLevel.Standard, WeaponLevel.Rifle);
        return true;
    }

    private bool TrainMilitary()
    {
        var hitotsu = _gameState.Planets.FirstOrDefault(p => p.Owner == FactionType.AI && p.Name == "Hitotsu");
        if (hitotsu == null)
            return false;

        var (equipment, weapon) = _difficulty switch
        {
            AIDifficulty.Easy => (EquipmentLevel.Standard, WeaponLevel.Pistol),
            AIDifficulty.Hard => (EquipmentLevel.Elite, WeaponLevel.Plasma),
            _ => (EquipmentLevel.Standard, WeaponLevel.Rifle)
        };

        var totalCost = PlatoonCosts.GetTotalCost(equipment, weapon);
        var cost = new ResourceCost { Credits = totalCost };
        if (!hitotsu.Resources.CanAfford(cost))
            return false;

        int troops = _random.Next(100, 151);
        _platoonSystem.CommissionPlatoon(hitotsu.ID, FactionType.AI, troops, equipment, weapon);
        return true;
    }

    private bool BuildEconomy()
    {
        bool mutated = false;

        foreach (var planet in _gameState.Planets.Where(p => p.Owner == FactionType.AI))
        {
            float buildChance = _personality == AIPersonality.Economic ? 0.8f : 0.4f;
            if (_random.NextDouble() > buildChance)
                continue;

            if (planet.Type == PlanetType.Volcanic)
            {
                _buildingSystem.StartConstruction(planet.ID, BuildingType.MiningStation);
                OnAIBuilding?.Invoke(planet.ID, BuildingType.MiningStation);
                mutated = true;
            }

            if (planet.Type == PlanetType.Tropical)
            {
                _buildingSystem.StartConstruction(planet.ID, BuildingType.HorticulturalStation);
                OnAIBuilding?.Invoke(planet.ID, BuildingType.HorticulturalStation);
                mutated = true;
            }
        }

        return mutated;
    }

    private bool LaunchAttack()
    {
        if (_personality == AIPersonality.Defensive && _random.NextDouble() > 0.2)
            return false;

        var target = _gameState.Planets
            .Where(p => p.Owner == FactionType.Player && p.Name != "Starbase")
            .OrderBy(p => GetPlanetDefenseStrength(p.ID))
            .FirstOrDefault();

        if (target == null)
            return false;

        var cruisers = _gameState.Craft
            .Where(c => c.Owner == FactionType.AI &&
                        c.Type == CraftType.BattleCruiser &&
                        c.CarriedPlatoonIDs.Count > 0)
            .ToList();

        if (cruisers.Count < 2)
            return false;

        OnAIAttacking?.Invoke(target.ID);
        return true;
    }

    private bool ExpandToNeutral()
    {
        var neutral = _gameState.Planets.FirstOrDefault(p => p.Owner == FactionType.Neutral);
        var hitotsu = _gameState.Planets.FirstOrDefault(p => p.Owner == FactionType.AI && p.Name == "Hitotsu");

        if (neutral == null || hitotsu == null)
            return false;

        _craftSystem.PurchaseCraft(CraftType.AtmosphereProcessor, hitotsu.ID, FactionType.AI);
        return true;
    }

    // ============================================================
    // 🔴 STOP-CONDITION FIX: always mutates state
    // ============================================================
    private void ForceGuaranteedMutation()
    {
        _gameState.AIFaction.AITurnNonce++;
    }

    private int CalculateMilitaryStrength(FactionType faction) { /* unchanged */ }
    private int ApplyDifficultyModifier(int strength) { /* unchanged */ }
    private bool IsUnderAttack() { /* unchanged */ }
    private bool CanAttack(float threatLevel) { /* unchanged */ }
    private int GetPlanetDefenseStrength(int planetID) { /* unchanged */ }
}

File: Overlord.Phaser/src/core/AIDecisionSystem.ts
import { GameState } from './GameState';
import { IncomeSystem } from './IncomeSystem';
import { ResourceSystem } from './ResourceSystem';
import { BuildingSystem } from './BuildingSystem';
import { CraftSystem } from './CraftSystem';
import { PlatoonSystem } from './PlatoonSystem';
import {
  FactionType,
  AIPersonality,
  AIDifficulty,
  PlanetType,
  BuildingType,
  BuildingStatus,
  CraftType,
  EquipmentLevel,
  WeaponLevel,
} from './models/Enums';
import { AIPersonalityConfig, AIAssessment } from './models/AIModels';
import { PlatoonCosts } from './models/PlatoonModels';
import { ResourceCost } from './models/ResourceModels';
import { BuildingCosts } from './models/BuildingModels';

/**
 * Platform-agnostic AI decision-making system.
 * Executes AI turn, evaluates game state, and makes strategic decisions.
 */
export class AIDecisionSystem {
  private readonly gameState: GameState;
  private readonly random: () => number;

  // System dependencies
  private readonly incomeSystem: IncomeSystem;
  private readonly resourceSystem: ResourceSystem;
  private readonly buildingSystem: BuildingSystem;
  private readonly craftSystem: CraftSystem;
  private readonly platoonSystem: PlatoonSystem;

  // AI configuration
  private personality: AIPersonality;
  private personalityConfig: AIPersonalityConfig;
  private difficulty: AIDifficulty;
  private didMutate: boolean = false;

  /**
   * Event fired when AI turn starts.
   */
  public onAITurnStarted?: () => void;

  /**
   * Event fired when AI turn completes.
   */
  public onAITurnCompleted?: () => void;

  /**
   * Event fired when AI initiates an attack.
   * Parameters: (targetPlanetID)
   */
  public onAIAttacking?: (targetPlanetID: number) => void;

  /**
   * Event fired when AI builds a structure.
   * Parameters: (planetID, buildingType)
   */
  public onAIBuilding?: (planetID: number, buildingType: BuildingType) => void;

  /**
   * Event fired when AI commissions a platoon.
   * Parameters: (planetID, troops, equipment, weapon)
   */
  public onAICommissioning?: (planetID: number, troops: number, equipment: EquipmentLevel, weapon: WeaponLevel) => void;

  /**
   * Event fired when AI purchases a craft.
   * Parameters: (craftType, planetID)
   */
  public onAIPurchasing?: (craftType: CraftType, planetID: number) => void;

  constructor(
    gameState: GameState,
    incomeSystem: IncomeSystem,
    resourceSystem: ResourceSystem,
    buildingSystem: BuildingSystem,
    craftSystem: CraftSystem,
    platoonSystem: PlatoonSystem,
    personality: AIPersonality = AIPersonality.Balanced,
    difficulty: AIDifficulty = AIDifficulty.Normal,
    randomFunc?: () => number,
  ) {
    if (!gameState) {
      throw new Error('gameState cannot be null or undefined');
    }
    if (!incomeSystem) {
      throw new Error('incomeSystem cannot be null or undefined');
    }
    if (!resourceSystem) {
      throw new Error('resourceSystem cannot be null or undefined');
    }
    if (!buildingSystem) {
      throw new Error('buildingSystem cannot be null or undefined');
    }
    if (!craftSystem) {
      throw new Error('craftSystem cannot be null or undefined');
    }
    if (!platoonSystem) {
      throw new Error('platoonSystem cannot be null or undefined');
    }

    this.gameState = gameState;
    this.incomeSystem = incomeSystem;
    this.resourceSystem = resourceSystem;
    this.buildingSystem = buildingSystem;
    this.craftSystem = craftSystem;
    this.platoonSystem = platoonSystem;

    this.personality = personality;
    this.personalityConfig = AIPersonalityConfig.getConfig(personality);
    this.difficulty = difficulty;
    this.random = randomFunc || Math.random;
  }

  /**
   * Gets the income system (for future AI income-based decisions).
   */
  public getIncomeSystem(): IncomeSystem {
    return this.incomeSystem;
  }

  /**
   * Gets the resource system (for future AI resource-based decisions).
   */
  public getResourceSystem(): ResourceSystem {
    return this.resourceSystem;
  }

  /**
   * Gets current AI personality.
   */
  public getPersonality(): AIPersonality {
    return this.personality;
  }

  /**
   * Gets AI personality name (e.g., "Commander Kratos").
   */
  public getPersonalityName(): string {
    return this.personalityConfig.name;
  }

  /**
   * Gets AI personality quote.
   */
  public getPersonalityQuote(): string {
    return this.personalityConfig.quote;
  }

  /**
   * Gets current difficulty level.
   */
  public getDifficulty(): AIDifficulty {
    return this.difficulty;
  }

  /**
   * Sets AI difficulty level.
   */
  public setDifficulty(difficulty: AIDifficulty): void {
    this.difficulty = difficulty;
  }

  /**
   * Executes full AI turn.
   */
  public executeAITurn(): void {
    this.onAITurnStarted?.();
    this.didMutate = false;
    const beforeFingerprint = this.createAIFingerprint();

    // Assess game state
    const assessment = this.assessGameState();

    // Execute decision tree
    this.executeDecisions(assessment);

    if (!this.didMutate) {
      this.applyFallbackAction();
    }

    const afterFingerprint = this.createAIFingerprint();
    if (beforeFingerprint === afterFingerprint) {
      console.error('[AI] No state change detected after AI turn', {
        turn: this.gameState.currentTurn,
        aiPlanets: this.gameState.planets.filter(p => p.owner === FactionType.AI).length,
        resources: {
          credits: this.gameState.aiFaction.resources.credits,
          minerals: this.gameState.aiFaction.resources.minerals,
          fuel: this.gameState.aiFaction.resources.fuel,
          food: this.gameState.aiFaction.resources.food,
          energy: this.gameState.aiFaction.resources.energy,
        },
      });
    }

    this.onAITurnCompleted?.();
  }

  /**
   * Assesses current game state (threat level, economy, territory).
   */
  public assessGameState(): AIAssessment {
    const assessment = new AIAssessment();

    // Calculate military strengths
    const playerMilitary = this.calculateMilitaryStrength(FactionType.Player);
    let aiMilitary = this.calculateMilitaryStrength(FactionType.AI);

    // Apply difficulty modifiers to AI strength
    aiMilitary = this.applyDifficultyModifier(aiMilitary);

    // Threat level = Player ÷ AI (higher = player stronger)
    assessment.threatLevel = aiMilitary > 0 ? playerMilitary / aiMilitary : 10.0;

    // Economic strength
    const playerRes = this.gameState.playerFaction.resources;
    const aiRes = this.gameState.aiFaction.resources;
    const playerTotal = playerRes.credits + playerRes.minerals + playerRes.fuel + playerRes.food;
    const aiTotal = aiRes.credits + aiRes.minerals + aiRes.fuel + aiRes.food;
    assessment.economicStrength = playerTotal > 0 ? aiTotal / playerTotal : 0.0;

    // Territory
    assessment.aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI).length;
    assessment.playerPlanets = this.gameState.planets.filter(p => p.owner === FactionType.Player).length;

    // Under attack check
    assessment.underAttack = this.isUnderAttack();

    // Can attack check
    assessment.canAttack = this.canAttack(assessment.threatLevel);

    return assessment;
  }

  /**
   * Executes AI decision tree based on assessment.
   * AI ALWAYS works toward endgame: Expand → Build Economy → Build Military → Attack
   */
  private executeDecisions(assessment: AIAssessment): void {
    console.log(`[AI] Turn ${this.gameState.currentTurn}: threatLevel=${assessment.threatLevel.toFixed(2)}, underAttack=${assessment.underAttack}, aiPlanets=${assessment.aiPlanets}, economicStrength=${assessment.economicStrength.toFixed(2)}`);

    // Priority 1: Defend if under attack (emergency response)
    if (assessment.underAttack) {
      console.log('[AI] Defending (under attack)');
      this.reinforceDefenses();
    }

    // Priority 2: ALWAYS expand to neutral planets (growth = victory)
    // Expansion is CRITICAL - more planets = more resources = more armies
    const neutralPlanets = this.gameState.planets.filter(p => p.owner === FactionType.Neutral);
    if (neutralPlanets.length > 0) {
      console.log('[AI] Expanding to neutral planets (ALWAYS)');
      this.expandToNeutral();
    }

    // Priority 3: ALWAYS build economy on owned planets (turn < 30 or Economic)
    // Need resources to build armies - build on ANY planet type
    if (this.gameState.currentTurn < 30 || this.personality === AIPersonality.Economic) {
      console.log('[AI] Building economy');
      this.buildEconomy();
    }

    // Priority 4: ALWAYS build military if we have resources
    // Build more aggressively if threatened
    if (assessment.threatLevel > 0.8 || this.gameState.currentTurn >= 5) {
      console.log('[AI] Building military');
      this.trainMilitary();
    }

    // Priority 5: Attack if we have military advantage
    // Personality affects aggressiveness, but all AIs attack eventually
    if (assessment.canAttack && !assessment.underAttack) {
      console.log('[AI] Launching attack');
      this.launchAttack();
    }
  }

  /**
   * Calculates total military strength for a faction.
   */
  private calculateMilitaryStrength(faction: FactionType): number {
    let totalStrength = 0;

    const platoons = this.gameState.platoons.filter(p => p.owner === faction);

    for (const platoon of platoons) {
      // Base strength from troop count
      let strength = platoon.troopCount;

      // Equipment multipliers
      switch (platoon.equipment) {
        case EquipmentLevel.Civilian:
          strength += 0;
          break;
        case EquipmentLevel.Standard:
          strength += Math.floor(platoon.troopCount / 2);
          break;
        case EquipmentLevel.Elite:
          strength += platoon.troopCount;
          break;
      }

      // Weapon multipliers
      switch (platoon.weapon) {
        case WeaponLevel.Pistol:
          strength += 0;
          break;
        case WeaponLevel.Rifle:
          strength += Math.floor(platoon.troopCount / 2);
          break;
        case WeaponLevel.Plasma:
          strength += platoon.troopCount;
          break;
      }

      totalStrength += strength;
    }

    return totalStrength;
  }

  /**
   * Applies difficulty modifier to AI military strength.
   */
  private applyDifficultyModifier(strength: number): number {
    switch (this.difficulty) {
      case AIDifficulty.Easy:
        return Math.floor(strength * 0.8); // -20%
      case AIDifficulty.Hard:
        return Math.floor(strength * 1.2); // +20%
      default:
        return strength; // Normal
    }
  }

  /**
   * Checks if AI is under attack.
   */
  private isUnderAttack(): boolean {
    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);

    for (const planet of aiPlanets) {
      // Check for player craft in orbit
      const playerCraft = this.gameState.craft.filter(
        c => c.planetID === planet.id && c.owner === FactionType.Player,
      );

      if (playerCraft.some(c => c.type === CraftType.BattleCruiser)) {
        return true; // Player Battle Cruiser at AI planet
      }
    }

    return false;
  }

  /**
   * Determines if AI can attack based on threat level and personality.
   */
  private canAttack(threatLevel: number): boolean {
    // Base attack thresholds by difficulty
    let baseThreshold: number;
    switch (this.difficulty) {
      case AIDifficulty.Easy:
        baseThreshold = 0.5; // AI needs 2× player strength
        break;
      case AIDifficulty.Hard:
        baseThreshold = 0.83; // AI needs 1.2× player strength
        break;
      default:
        baseThreshold = 0.67; // AI needs 1.5× player strength (Normal)
    }

    // Apply personality aggression modifier
    const adjustedThreshold = baseThreshold + this.personalityConfig.aggressionModifier;

    // AI can attack if threat level is below threshold (AI is stronger)
    return threatLevel < adjustedThreshold;
  }

  /**
   * Reinforces defenses at threatened planets.
   */
  private reinforceDefenses(): void {
    const hitotsu = this.gameState.planets.find(p => p.owner === FactionType.AI && p.name === 'Hitotsu');
    if (!hitotsu) {
      return;
    }

    // Check if we can afford a platoon
    const totalCost = PlatoonCosts.getTotalCost(EquipmentLevel.Standard, WeaponLevel.Rifle);
    const cost = new ResourceCost();
    cost.credits = totalCost;
    if (!hitotsu.resources.canAfford(cost)) {
      return;
    }

    // Train defensive platoon
    const platoonId = this.platoonSystem.commissionPlatoon(
      hitotsu.id,
      FactionType.AI,
      100,
      EquipmentLevel.Standard,
      WeaponLevel.Rifle,
    );
    if (platoonId >= 0) {
      this.didMutate = true;
    }
    this.onAICommissioning?.(hitotsu.id, 100, EquipmentLevel.Standard, WeaponLevel.Rifle);
  }

  /**
   * Trains military units on ANY AI planet that can afford it.
   */
  private trainMilitary(): void {
    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);
    console.log(`[AI] trainMilitary: AI owns ${aiPlanets.length} planets`);

    if (aiPlanets.length === 0) {
      console.log('[AI]   No AI planets to train on!');
      return;
    }

    // Determine equipment/weapons by difficulty
    let equipment: EquipmentLevel;
    let weapon: WeaponLevel;

    switch (this.difficulty) {
      case AIDifficulty.Easy:
        equipment = EquipmentLevel.Standard;
        weapon = WeaponLevel.Pistol;
        break;
      case AIDifficulty.Hard:
        equipment = EquipmentLevel.Elite;
        weapon = WeaponLevel.Plasma;
        break;
      default:
        equipment = EquipmentLevel.Standard;
        weapon = WeaponLevel.Rifle; // Normal
    }

    // Check if we can afford
    const totalCost = PlatoonCosts.getTotalCost(equipment, weapon);
    const cost = new ResourceCost();
    cost.credits = totalCost;

    // Try to train on any planet that can afford it
    for (const planet of aiPlanets) {
      console.log(`[AI]   Checking ${planet.name} for platoon commission`);

      if (!planet.resources.canAfford(cost)) {
        console.log(`[AI]     Cannot afford (need ${totalCost} credits, have ${planet.resources.credits})`);
        continue;
      }

      // Train platoon
      const troops = Math.floor(this.random() * 51) + 100; // 100-150
      console.log(`[AI]     Commissioning ${troops} troops with ${EquipmentLevel[equipment]}/${WeaponLevel[weapon]}`);
      const platoonId = this.platoonSystem.commissionPlatoon(planet.id, FactionType.AI, troops, equipment, weapon);

      if (platoonId >= 0) {
        console.log(`[AI]     ✓ Platoon ${platoonId} commissioned on ${planet.name}`);
        this.onAICommissioning?.(planet.id, troops, equipment, weapon);
        this.didMutate = true;
        return; // Train one per turn
      } else {
        console.log(`[AI]     ✗ Commission failed`);
      }
    }

    console.log('[AI]   No planets could afford military training');
  }

  /**
   * Builds economic infrastructure on ANY planet type.
   * Priority: Mining > Horticultural > Docking Bay (for craft production)
   */
  private buildEconomy(): void {
    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);
    console.log(`[AI] buildEconomy: AI owns ${aiPlanets.length} planets`);

    for (const planet of aiPlanets) {
      console.log(`[AI]   Checking planet ${planet.name} (${PlanetType[planet.type]})`);

      // Economic personality builds more aggressively (80% vs 60%)
      const buildChance = this.personality === AIPersonality.Economic ? 0.8 : 0.6;
      const roll = this.random();
      if (roll > buildChance) {
        console.log(`[AI]     Skip (roll ${roll.toFixed(2)} > ${buildChance})`);
        continue;
      }

      // Try to build in priority order: Mining → Horticultural → Docking Bay
      // Build whatever we can afford and have slots for
      let built = false;

      // Priority 1: Mining Station (minerals + fuel)
      if (!built && this.buildingSystem.canBuild(planet.id, BuildingType.MiningStation)) {
        console.log(`[AI]     Building Mining Station`);
        const result = this.buildingSystem.startConstruction(planet.id, BuildingType.MiningStation);
        console.log(`[AI]     Build result: ${result}`);
        if (result) {
          this.onAIBuilding?.(planet.id, BuildingType.MiningStation);
          this.didMutate = true;
          built = true;
        }
      }

      // Priority 2: Horticultural Station (food for population growth)
      if (!built && this.buildingSystem.canBuild(planet.id, BuildingType.HorticulturalStation)) {
        console.log(`[AI]     Building Horticultural Station`);
        const result = this.buildingSystem.startConstruction(planet.id, BuildingType.HorticulturalStation);
        console.log(`[AI]     Build result: ${result}`);
        if (result) {
          this.onAIBuilding?.(planet.id, BuildingType.HorticulturalStation);
          this.didMutate = true;
          built = true;
        }
      }

      // Priority 3: Docking Bay (enables craft production)
      if (!built && this.buildingSystem.canBuild(planet.id, BuildingType.DockingBay)) {
        console.log(`[AI]     Building Docking Bay`);
        const result = this.buildingSystem.startConstruction(planet.id, BuildingType.DockingBay);
        console.log(`[AI]     Build result: ${result}`);
        if (result) {
          this.onAIBuilding?.(planet.id, BuildingType.DockingBay);
          this.didMutate = true;
          built = true;
        }
      }

      if (!built) {
        console.log(`[AI]     No valid builds available (slots full or no resources)`);
      }
    }
  }

  /**
   * Launches attack on player planet.
   */
  private launchAttack(): void {
    // Aggressive personality attacks more often
    if (this.personality === AIPersonality.Defensive) {
      // Defensive personality rarely attacks
      if (this.random() > 0.2) {
        return;
      }
    }

    // Find target (weakest player planet)
    const playerPlanets = this.gameState.planets.filter(p => p.owner === FactionType.Player);
    if (playerPlanets.length === 0) {
      return;
    }

    // Avoid Starbase initially
    let target = playerPlanets
      .filter(p => p.name !== 'Starbase')
      .sort((a, b) => this.getPlanetDefenseStrength(a.id) - this.getPlanetDefenseStrength(b.id))[0];

    if (!target) {
      target = playerPlanets[0]; // Attack Starbase if only option
    }

    // Get AI Battle Cruisers with platoons
    const battleCruisers = this.gameState.craft.filter(
      c => c.owner === FactionType.AI && c.type === CraftType.BattleCruiser && c.carriedPlatoonIDs.length > 0,
    );

    if (battleCruisers.length < 2) {
      return; // Need at least 2 cruisers
    }

    // NOTE: Actual movement would require a NavigationSystem
    // For now, just fire the event
    this.onAIAttacking?.(target.id);
  }

  /**
   * Expands to neutral planets by purchasing Atmosphere Processors.
   */
  private expandToNeutral(): void {
    const neutralPlanets = this.gameState.planets.filter(p => p.owner === FactionType.Neutral);
    console.log(`[AI] expandToNeutral: ${neutralPlanets.length} neutral planets available`);

    if (neutralPlanets.length === 0) {
      console.log('[AI]   No neutral planets to expand to');
      return;
    }

    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);
    if (aiPlanets.length === 0) {
      console.log('[AI]   No AI planets to purchase from!');
      return;
    }

    // Try to purchase Atmosphere Processor from any AI planet
    for (const planet of aiPlanets) {
      console.log(`[AI]   Trying to purchase Atmosphere Processor from ${planet.name}`);

      const craftId = this.craftSystem.purchaseCraft(CraftType.AtmosphereProcessor, planet.id, FactionType.AI);

      if (craftId >= 0) {
        console.log(`[AI]     ✓ Atmosphere Processor ${craftId} purchased!`);
        this.onAIPurchasing?.(CraftType.AtmosphereProcessor, planet.id);
        this.didMutate = true;
        return; // Purchase one per turn
      } else {
        console.log(`[AI]     ✗ Purchase failed (no resources or no docking bay)`);
      }
    }

    console.log('[AI]   Could not purchase Atmosphere Processor from any planet');
  }

  /**
   * Gets defense strength of a planet (garrison + defenses).
   */
  private getPlanetDefenseStrength(planetID: number): number {
    let strength = 0;

    // Garrison strength
    const garrison = this.gameState.platoons.filter(p => p.planetID === planetID);
    for (const platoon of garrison) {
      strength += platoon.troopCount;
    }

    // Orbital Defense bonus
    const planet = this.gameState.planetLookup.get(planetID);
    if (planet) {
      const hasOrbitalDefense = planet.structures.some(
        s => s.type === BuildingType.OrbitalDefense && s.status === BuildingStatus.Active,
      );
      if (hasOrbitalDefense) {
        strength = Math.floor(strength * 1.2); // +20% defense bonus
      }
    }

    return strength;
  }

  private applyFallbackAction(): void {
    if (this.adjustTaxRateFallback()) {
      return;
    }

    if (this.queueCheapestConstruction()) {
      return;
    }

    if (this.commissionCheapestPlatoon()) {
      return;
    }

    console.warn('[AI] No fallback action available');
  }

  private adjustTaxRateFallback(): boolean {
    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);
    if (aiPlanets.length === 0) {
      return false;
    }

    for (const planet of aiPlanets) {
      const currentRate = planet.taxRate;
      let nextRate = currentRate;

      if (currentRate < 50) {
        nextRate = currentRate + 1;
      } else if (currentRate > 50) {
        nextRate = currentRate - 1;
      } else if (currentRate < 100) {
        nextRate = currentRate + 1;
      } else if (currentRate > 0) {
        nextRate = currentRate - 1;
      }

      nextRate = Math.max(0, Math.min(100, nextRate));

      if (nextRate !== currentRate) {
        planet.taxRate = nextRate;
        this.didMutate = true;
        return true;
      }
    }

    return false;
  }

  private queueCheapestConstruction(): boolean {
    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);
    if (aiPlanets.length === 0) {
      return false;
    }

    const buildOptions = [
      BuildingType.MiningStation,
      BuildingType.HorticulturalStation,
      BuildingType.DockingBay,
      BuildingType.OrbitalDefense,
      BuildingType.SurfacePlatform,
    ];

    for (const planet of aiPlanets) {
      const affordableOptions = buildOptions
        .filter(type => this.buildingSystem.canBuild(planet.id, type))
        .sort((a, b) => {
          const costA = BuildingCosts.getCost(a);
          const costB = BuildingCosts.getCost(b);
          return (costA.credits + costA.minerals + costA.fuel + costA.food + costA.energy)
            - (costB.credits + costB.minerals + costB.fuel + costB.food + costB.energy);
        });

      for (const buildingType of affordableOptions) {
        if (this.buildingSystem.startConstruction(planet.id, buildingType)) {
          this.onAIBuilding?.(planet.id, buildingType);
          this.didMutate = true;
          return true;
        }
      }
    }

    return false;
  }

  private commissionCheapestPlatoon(): boolean {
    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);
    if (aiPlanets.length === 0) {
      return false;
    }

    const equipment = EquipmentLevel.Standard;
    const weapon = WeaponLevel.Pistol;
    const totalCost = PlatoonCosts.getTotalCost(equipment, weapon);

    for (const planet of aiPlanets) {
      if (planet.resources.credits < totalCost) {
        continue;
      }
      if (planet.population < PlatoonSystem.MinTroops) {
        continue;
      }

      const platoonId = this.platoonSystem.commissionPlatoon(
        planet.id,
        FactionType.AI,
        PlatoonSystem.MinTroops,
        equipment,
        weapon,
      );

      if (platoonId >= 0) {
        this.onAICommissioning?.(planet.id, PlatoonSystem.MinTroops, equipment, weapon);
        this.didMutate = true;
        return true;
      }
    }

    return false;
  }

  private createAIFingerprint(): string {
    const resources = this.gameState.aiFaction.resources;
    const aiPlanets = this.gameState.planets.filter(p => p.owner === FactionType.AI);
    const taxRates = aiPlanets.map(p => p.taxRate).join(',');
    const queuedBuilds = aiPlanets.reduce(
      (count, planet) => count + planet.structures.filter(s => s.status === BuildingStatus.UnderConstruction).length,
      0,
    );
    const fleetOrders = this.gameState.craft.filter(
      c => c.owner === FactionType.AI && c.inTransit,
    ).length;

    return JSON.stringify({
      resources: {
        credits: resources.credits,
        minerals: resources.minerals,
        fuel: resources.fuel,
        food: resources.food,
        energy: resources.energy,
      },
      taxRates,
      queuedBuilds,
      fleetOrders,
    });
  }
}
